crates/burn-jit/src/kernel/matmul/tiling2d_shader.rs

```diff
@@ -0,0 +1,636 @@
+        let rank = Variable::Rank;
+        let ultimate_dim = scope.create_local(Elem::UInt);
+        let penultimate_dim = scope.create_local(Elem::UInt);
+        gpu!(scope, ultimate_dim = rank - 1u32);
+        gpu!(scope, penultimate_dim = rank - 2u32);
+        let dim_m = scope.create_local(Elem::UInt);
+        let dim_k = scope.create_local(Elem::UInt);
+        let dim_n = scope.create_local(Elem::UInt);
+        gpu!(scope, dim_m = shape(lhs, penultimate_dim));
+        gpu!(scope, dim_k = shape(lhs, ultimate_dim));
+        gpu!(scope, dim_n = shape(rhs, ultimate_dim));
+
+        // Strides
+        let lhs_stride_row = scope.create_local(Elem::UInt);
+        let lhs_stride_col = scope.create_local(Elem::UInt);
+        let rhs_stride_row = scope.create_local(Elem::UInt);
+        let rhs_stride_col = scope.create_local(Elem::UInt);
+        let out_stride_row = scope.create_local(Elem::UInt);
+        let out_stride_col = scope.create_local(Elem::UInt);
+        gpu!(scope, lhs_stride_row = stride(lhs, penultimate_dim));
+        gpu!(scope, lhs_stride_col = stride(lhs, ultimate_dim));
+        gpu!(scope, rhs_stride_row = stride(rhs, penultimate_dim));
+        gpu!(scope, rhs_stride_col = stride(rhs, ultimate_dim));
+        gpu!(scope, out_stride_row = stride(out, penultimate_dim));
+        gpu!(scope, out_stride_col = stride(out, ultimate_dim));
+
+        // Workgroup offset
+        let skip_row = scope.create_local(Elem::UInt);
+        let workgroup_id_x = Variable::WorkgroupIdX;
+        gpu!(scope, skip_row = workgroup_id_x);
+        gpu!(scope, skip_row *= block_size_m);
+        let skip_col = scope.create_local(Elem::UInt);
+        let workgroup_id_y = Variable::WorkgroupIdY;
+        gpu!(scope, skip_col = workgroup_id_y);
+        gpu!(scope, skip_col *= block_size_n);
+
+        // Position of the first element of the thread, relative to the block
+        let thread_row = scope.create_local(Elem::UInt);
+        gpu!(scope, thread_row = local_idx / n_threads_per_row);
+        gpu!(scope, thread_row *= tile_size_m);
+        let thread_col = scope.create_local(Elem::UInt);
+        gpu!(scope, thread_col = local_idx % n_threads_per_row);
+        gpu!(scope, thread_col *= tile_size_n);
+
+        // Position of the first element of the thread, in absolute (in one batch)
+        let row = scope.create_local(Elem::UInt);
+        let col = scope.create_local(Elem::UInt);
+        gpu!(scope, row = skip_row + thread_row);
+        gpu!(scope, col = skip_col + thread_col);
+
+        // Calculate offset.
+        let offset_lhs = scope.create_local(Elem::UInt);
+        let offset_rhs = scope.create_local(Elem::UInt);
+        gpu!(scope, offset_lhs = skip_row * lhs_stride_row);
+        gpu!(scope, offset_rhs = skip_col * rhs_stride_col);
+
+        // Batch offset for the output.
+        let offset_output = scope.create_local(Elem::UInt);
+        let batch_dims = scope.create_local(Elem::UInt);
```

Comment:
I would create a function with almost everything at the beginning of the shader that returns a struct with all matrix multiply information required.

1. Gather the kernel information (offset, k, m, n, etc.)
2. Read the lhs & rhs tensor and fill the shared memory, then return a struct with the result.
3. Perform the matmul calculation and fill the output shared memory, then return a struct with the result.
4. Write the output shared memory to the output.